app/questionnaire/questionnaire_store_updater.py

```diff
@@ -159,27 +161,30 @@ def remove_list_item_data(self, list_name: str, list_item_id: str) -> None:
             self._update_relationship_question_completeness(list_name)
 
         self._progress_store.remove_progress_for_list_item_id(list_item_id=list_item_id)
-        self._capture_dependencies_for_list_item_removal(list_name)
+        self._capture_dependencies_for_list_change(list_name)
 
-    def _capture_dependencies_for_list_item_removal(self, list_name: str) -> None:
+    def _capture_dependencies_for_list_change(self, list_name: str) -> None:
         """
-        If an item is removed from a list, dependencies of any child blocks need to be captured and their progress updated.
-        (E.g. dynamic-answers depending on a child block could have validation, so need revisiting if an item is removed)
+        Captures the dependencies when an item is added to or removed from the given list.
+        Any list collector sections will be affected by the change as well as other sections using when rules
         """
-        for list_collector in self._schema.get_list_collectors_for_list(
-            for_list=list_name
-        ):
-            child_blocks = (
-                list_collector.get("add_block"),
-                list_collector.get("remove_block"),
-                *list_collector.get("repeating_blocks", []),
-            )
-            for child_block in child_blocks:
-                if child_block:
-                    for answer_id in self._schema.get_answer_ids_for_block(
-                        child_block["id"]
-                    ):
-                        self._capture_block_dependencies_for_answer(answer_id)
+        self._capture_block_dependencies_for_list(list_name)
+        section_ids = self._schema.get_when_rule_section_dependencies_for_list(
+            list_name
+        )
+        section_ids.update(
+            self._schema.list_collector_section_ids_by_list_name[list_name]
+        )
+
+        for section_key in self.started_section_keys(section_ids=section_ids):
+            # Only add sections which are repeated sections for this list, or the section in which this list is collected
+            # Prevents list item progresses being added as dependants as these are captured by started_section_keys(section_ids=section_ids)
+            if (
+                not self._schema.get_repeat_for_section(section_key.section_id)
+                and section_key.list_item_id
```

Comment:
Minor, would be more performant to flip this logic, check if we have list item id, if so, fetch repeat list name.